### ISSUE TITLE
Nested rows bad state after changing data

### ISSUE THREAD

**Issue by aaronbeall** [handsontable/handsontable#5753]
### Description

Because https://github.com/handsontable/handsontable/issues/3919 makes sorting incompatible with the Nested Rows plugin I have attempted to implement my own custom sorting following the example in [`setSortConfig()`](https://handsontable.com/docs/6.2.2/ColumnSorting.html#setSortConfig). The problem is that after updating the source data, rendering gets messed up, especially if any rows are collapsed at any point. I have tried a lot of things (listed below) but no matter what I cannot get the table to re-render in the correct state. The source data is correct, but the table is rendering wrong rows.

### Steps to reproduce

1. Open the example here: https://jsfiddle.net/9fspzoew/
2. Collapse the first row
3. Sort on "col1" -- this is where a custom sort and updating the data happens
4. Expand the first row

Result is that "Child" row is rendered all over the table (this is wrong):
![](http://g.recordit.co/vRmwU2q9cu.gif)

This is just one example. I've run into many similar issues, both with and without actually collapsing any rows. In all of the issues the data just won't render correctly after updating it. I've tried:
- Updating by directly mutating the source `data`.
- Updating by calling `updateSettings({ data: newData })`.
- Updating by using `loadData()` with a new array.
- Re-rendering after updating by calling `render()`.
- Directly clearing the nested rows cache using `hot.getPlugin("nestedRows").dataManager.rewriteCache()`.
- Triggering redraw after updating by calling `updateSettings({ width: 500 })` -- this actually works the best out of everything I've tried, but not without issues as seen above.
- Resetting nested rows by turning it off, loading a blank array, setting a timeout, turning it on and loading the sorted array. Same issues.
- Setting `observeChanges: false` in case that was causing issues. No difference.
- Disabled `columnSorting` altogether (commented out code in demo applies update with sorted data using a custom button instead of table sort action in header). Same issues.
- Cloning all data in case an internal map is stumbling over all the moved row objects. No difference.

The `sortTable()` function in the demo is the relevant part.

No matter what, after updating data something ends up rendering wrong related to row nesting. If I never change the data it works fine. Some kind of internal cache is getting into a bad state after updating, it seems. The only thing that works is to completely destroy and recreate the entire table instance, but this also causes column state (sort indicator, column sizes, etc) and table scroll position to be lost, which is not acceptable. 

**I need help forcing the table to redraw correctly.** I don't care if its relatively slow, it's only needed when a user sorts. I don't care if it resets all expand rows state. I don't care if it resets scroll position to top, but I need to maintain horizontal scroll. I don't care if it's hacky and involves reaching deep into core code and busting some cache I'm not supposed to know about... I just need it to render correctly. :)

### Demo

https://jsfiddle.net/9fspzoew/

### Your environment
* Handsontable version: 6.2.2 Pro
* Browser Name and version: Chrome 71.0.3578.98
* Operating System: MacOS Mojave 10.14.1

**Comment by aaronbeall** [comment]
Messed with this a bunch in my app and eventually found out that a lot of problems go away in this specific demo if I just set the sort comparator to always return 0.
```
  columnSorting: {
    // ...
    compareFunctionFactory: () => () => 0
  }
```
Fixed demo: https://jsfiddle.net/ymz7arwv/

I was already blocking default sort by returning false in `beforeColumnSort` but this was needed to prevent things rendering weird.

I also noticed that even if I completely re-instantiated the table with new data, after simply using `setSortConfig()` *before* I made the above change, things would still get into a bad state. But once I added the `() => 0` comparator things started working. Not sure if that helps identify the root issue...

**Comment by AMBudnik** [comment]
No idea why I did not spot this issue before. Have you done further research, Aaron?

**Comment by aaronbeall** [comment]
@AMBudnik Not really much to report. At my last post I was able to sort data (outside the table) without rebuilding the table (using the explicit `compareFunctionFactory`) but I ran into a separate problem with nested rows when implementing custom filtering (filtering data outside the table), basically this issue: https://github.com/handsontable/handsontable/issues/3914 where trying to render 0 rows I would give an error, or after rendering 0 rows then trying to render some rows would give an error. Similar error in nested row plugin about row mapping. I had to go back to re-instantiating the table any time rows change. Looking forward to the data mapping refactor, crossing my fingers that helps my case.

**Comment by AMBudnik** [comment]
>  I'm looking forward to the data mapping refactor (#5112)

We're currently working on it https://github.com/handsontable/handsontable/issues/5751
I also think that this will fix most of the issues related to index translation failures.

**Comment by AMBudnik** [comment]
Hey @aaronbeall 

we finally can test the `8.0.0-rc1` branch, ale it seems to change the behavior https://jsfiddle.net/m9a67jsw/ but the order of columns seem to be random

**Comment by aaronbeall** [comment]
@AMBudnik This looks really promising! What do you mean by the "the order of columns seem to be random"?

**Comment by AMBudnik** [comment]
Typo, it should be `rows`, not `columns`.
We'll be having some new builds soon. Today we've released 7.2.0 without the index mapper changes.

**Comment by aaronbeall** [comment]
Hm. The order actually looked correct to me. The only thing strange is that clicking a row collapse button would automatically expand the previously collapsed row (even without any sorting taking place).

![](http://g.recordit.co/4e2qqccYSc.gif)

What is the index mapper issue exactly?

**Comment by wojciechczerniak** [comment]
@aaronbeall This was already fixed in the latest build https://jsfiddle.net/wgbce4Ly/

**Comment by AMBudnik** [comment]
Remains solved in 8.0.0.-beta2-rev6 https://jsfiddle.net/sp3d20u7/

**Comment by AMBudnik** [comment]
**Works well** in 8.0.0-beta2-rev16 https://jsfiddle.net/AMBudnik/3w87rz2q/

**Comment by pnowak** [comment]
Hi @aaronbeall 

I have good news! We've just released 8.0.0-beta2 that fixes the mentioned issue. We won't be closing this ticket now. It will be closed after we release the final 8.0.0 version of Handsontable.

Here’s a full list of changes https://github.com/handsontable/handsontable/releases/tag/8.0.0-beta.2
NPM https://www.npmjs.com/package/handsontable/v/8.0.0-beta.2
 
All feedback's appreciated. Thank you for your input.

**Comment by AMBudnik** [comment]
@aaronbeall we can officially close this issue as solved in 8.0.0 :-) 🥇  Thank you for your contribution.
